src/gennylib/src/encryption.cpp

```diff
@@ -0,0 +1,527 @@
+    }
+
+private:
+    EncryptedCollectionMap<FLEEncryptedCollection> _fleCollections;
+    std::string _keyVaultDb;
+    std::string _keyVaultColl;
+};
+
+}  // namespace genny::v1
+
+namespace {
+
+/**
+ * This struct represents a node in the tree-like structure
+ * of an FLE schema.
+ */
+struct FLEFieldPathNode {
+    std::string name;
+    std::vector<FLEFieldPathNode> subnodes;
+    const genny::v1::EncryptedField* field;
+
+    /**
+     * Helper function for building the FLE schema document, which has a
+     * hierarchical structure.
+     *
+     * @param subdoc the document builder to append fields onto
+     * @param root whether this node is the root, which does not have a name
+     */
+    void append(bsoncxx::builder::basic::sub_document subdoc, bool root = false) const {
+        using bsoncxx::builder::basic::kvp;
+        using bsoncxx::builder::basic::sub_document;
+
+        auto appendProperties = [](sub_document fieldDoc,
+                                   const std::vector<FLEFieldPathNode>& childNodes) {
+            fieldDoc.append(kvp("properties", [&](sub_document propDoc) {
+                for (auto& n : childNodes) {
+                    n.append(propDoc);
+                }
+            }));
+            fieldDoc.append(kvp("bsonType", "object"));
+        };
+
+        if (subnodes.empty()) {
+            subdoc.append(kvp(name, [&](sub_document fieldDoc) {
+                fieldDoc.append(kvp("encrypt", [&](sub_document encryptDoc) {
+                    field->appendEncryptInfo(encryptDoc);
+                }));
+            }));
+        } else if (!root) {
+            subdoc.append(
+                kvp(name, [&](sub_document fieldDoc) { appendProperties(fieldDoc, subnodes); }));
+        } else {
+            appendProperties(subdoc, subnodes);
+        }
+    }
+};
+
+/**
+ * Splits a dot-separated path string into its constituent parts &
+ * returns a list.
+ *
+ * @param path the dotted path string
+ * @return a non-empty ordered list of parsed path components
+ * @throws InvalidConfigurationException
+ *   if path is empty, or has empty path segments
+ */
+std::vector<std::string> splitDottedPath(const std::string& path) {
+    std::vector<std::string> parts;
+    boost::split(parts, path, boost::is_any_of("."));
+    if (std::any_of(parts.begin(), parts.end(), [](auto& part) { return part.empty(); })) {
+        std::ostringstream ss;
+        ss << "Field path \"" << path << "\" is not a valid path";
+        throw genny::InvalidConfigurationException(ss.str());
+    }
+    return parts;
+}
+
+}  // namespace
+
+namespace genny::v1 {
+
+using bsoncxx::builder::basic::kvp;
+using bsoncxx::builder::basic::make_document;
+using bsoncxx::builder::basic::sub_array;
+using bsoncxx::builder::basic::sub_document;
+using bsoncxx::types::bson_value::view_or_value;
+
+EncryptedField::EncryptedField(const Node& yaml) {
+    _path = yaml.key();
+    _type = yaml["type"].to<std::string>();
```

Comment:
what happens if type is missing?

Reply:
it will throw an exception, with an error saying `"Invalid key 'type': Tried to access node that doesn't exist. On node with path ..."`